File: be/src/main/java/sg/flow/entities/TransactionHistory.java
package sg.flow.entities;

import java.time.LocalDate;
import java.time.LocalTime;

import org.springframework.data.annotation.Id;
import org.springframework.data.relational.core.mapping.Table;
import org.springframework.lang.Nullable;

import jakarta.validation.constraints.NotNull;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Builder.Default;
import lombok.Data;

@Data
@AllArgsConstructor
@Builder
@Table(name = "transaction_history")
public class TransactionHistory {

    @Id
    private Long id;

    @NotNull
    private String transactionReference;

    @Nullable
    private Account account;

    @Nullable
    private Card card;

    @NotNull
    private LocalDate transactionDate;

    @Nullable
    private LocalTime transactionTime;

    @NotNull
    private Double amount;

    @NotNull
    private String transactionType;

    @Default
    private String description = "";

    @NotNull
    private String transactionStatus;

    @Default
    private String friendlyDescription = "";
}

File: be/src/main/kotlin/sg/flow/entities/TransactionHistory.kt
package sg.flow.entities

import jakarta.validation.constraints.NotBlank
import jakarta.validation.constraints.NotNull
import java.time.LocalDate
import java.time.LocalTime
import org.springframework.data.annotation.Id
import sg.flow.models.card.BriefCard

data class TransactionHistory(
        @field:Id val id: Long?,
        @field:NotBlank val transactionReference: String,
        val account: Account? = null,
        val card: BriefCard? = null,
        @field:NotNull val transactionDate: LocalDate,
        val transactionTime: LocalTime? = null,
        @field:NotNull val amount: Double,
        @field:NotBlank val transactionType: String,
        val description: String = "",
        @field:NotBlank val transactionStatus: String,
        val friendlyDescription: String = "",
        val transactionCategory: String = "",
        val extractedCardNumber: String? = null,
        val revisedTransactionDate: LocalDate? = null,
        val brandName: String? = null,
        val isProcessed: Boolean = false,
        val isIncludedInSpendingOrIncome: Boolean = true,
        val finverseId: String,
)
